File: app-android/app/src/main/java/com/goexercise/app/domain/StreakCalculator.kt
package com.goexercise.app.domain

import java.time.LocalDate

/**
 * 連続記録の計算。iOS `StreakCalculator.swift` の 1:1 移植。
 * 休息日(Rest)は連続記録を**途切れさせずスキップ**する(カウントもしない)。
 * rescuedDates は AchievementEvaluator 経由で achieved 扱いになる。
 */
object StreakCalculator {

    fun currentStreak(
        records: List<WorkoutRecord>,
        today: LocalDate,
        rescuedDates: Set<LocalDate> = emptySet(),
        restLimit: Int = 2,
    ): Int {
        var streak = 0
        var cursor = today

        while (true) {
            val restDays = RestDayResolver.restDaySet(cursor, records, today, restLimit)
            val status = AchievementEvaluator.dailyStatus(
                date = cursor,
                records = records,
                restDays = restDays,
                rescuedDates = rescuedDates,
                today = today,
            )

            when (status) {
                // Rescued(保険チケット救済)も達成同等に連続へ加算(漏らすとフリーズ日で連続が切れる)。
                DailyStatus.Achieved, DailyStatus.Rescued, DailyStatus.TodayAchieved -> streak += 1
                // Rest = カウントしないが連続も切れない。
                // TodayPending = 今日まだ未記録 → 連続を切らずスキップし「昨日までの連続」を数える
                // (今日記録すれば TodayAchieved になり加算される。iOS StreakCalculator と同一挙動)。
                DailyStatus.Rest, DailyStatus.TodayPending -> Unit // skip
                else -> return streak
            }

            cursor = cursor.minusDays(1)
        }
    }

    fun streakState(
        records: List<WorkoutRecord>,
        today: LocalDate,
        rescuedDates: Set<LocalDate> = emptySet(),
        lookbackDays: Long = 365,
        restLimit: Int = 2,
    ): StreakState {
        val current = currentStreak(records, today, rescuedDates, restLimit)
        var longest = 0
        var running = 0
        var lastAchievedDate: LocalDate? = null

        var cursor = today.minusDays(lookbackDays)
        while (!cursor.isAfter(today)) {
            val restDays = RestDayResolver.restDaySet(cursor, records, today, restLimit)
            val status = AchievementEvaluator.dailyStatus(
                date = cursor,
                records = records,
                restDays = restDays,
                rescuedDates = rescuedDates,
                today = today,
            )

            when (status) {
                DailyStatus.Achieved, DailyStatus.TodayAchieved -> {
                    running += 1
                    longest = maxOf(longest, running)
                    lastAchievedDate = cursor
                }
                // Rest と TodayPending は running を保持して連続を切らない
                // (今日未記録でも昨日までの連続を維持。iOS streakState と同一挙動)。
                DailyStatus.Rest, DailyStatus.TodayPending -> Unit
                else -> running = 0
            }

            cursor = cursor.plusDays(1)
        }

        return StreakState(currentStreak = current, longestStreak = longest, lastAchievedDate = lastAchievedDate)
    }
}

File: app/GOExercise/GOExercise/Services/StreakCalculator.swift
import Foundation

enum StreakCalculator {
    static func currentStreak(
        records: [WorkoutRecord],
        today: Date,
        rescuedDates: Set<Date> = [],
        restLimit: Int = 2,
        calendar: Calendar = .mondayFirst
    ) -> Int {
        var streak = 0
        var cursor = calendar.startOfDay(for: today)

        while true {
            let restDays = RestDayResolver.restDaySet(
                for: cursor,
                records: records,
                today: today,
                limit: restLimit,
                calendar: calendar
            )
            let status = AchievementEvaluator.dailyStatus(
                for: cursor,
                records: records,
                restDays: restDays,
                rescuedDates: rescuedDates,
                today: today,
                calendar: calendar
            )

            switch status {
            case .achieved, .todayAchieved:
                streak += 1
            case .rest, .todayPending:
                // rest = 連続を切らずスキップ。
                // todayPending = 今日まだ未記録 → 連続を切らずスキップし「昨日までの連続」を数える。
                // (今日記録すれば .todayAchieved になり今日分も加算される。フリーズ復活直後も
                //  昨日までの連続が即表示されるようにするための基準。todayPending は今日のみ発生。)
                break  // skip — does not count, but does not break the streak
            default:
                return streak
            }

            guard let previous = calendar.date(byAdding: .day, value: -1, to: cursor) else {
                return streak
            }
            cursor = previous
        }
    }

    static func streakState(
        records: [WorkoutRecord],
        today: Date,
        rescuedDates: Set<Date> = [],
        lookbackDays: Int = 365,
        restLimit: Int = 2,
        calendar: Calendar = .mondayFirst
    ) -> StreakState {
        let todayStart = calendar.startOfDay(for: today)
        let current = currentStreak(
            records: records, today: todayStart,
            rescuedDates: rescuedDates, restLimit: restLimit, calendar: calendar
        )
        var longest = 0
        var running = 0
        var lastAchievedDate: Date?

        let start = calendar.date(byAdding: .day, value: -lookbackDays, to: todayStart) ?? todayStart
        var cursor = start

        while cursor <= todayStart {
            let restDays = RestDayResolver.restDaySet(
                for: cursor,
                records: records,
                today: todayStart,
                limit: restLimit,
                calendar: calendar
            )
            let status = AchievementEvaluator.dailyStatus(
                for: cursor,
                records: records,
                restDays: restDays,
                rescuedDates: rescuedDates,
                today: todayStart,
                calendar: calendar
            )

            switch status {
            case .achieved, .todayAchieved:
                running += 1
                longest = max(longest, running)
                lastAchievedDate = cursor
            case .rest:
                break  // skip — running is preserved
            default:
                running = 0
            }

            guard let next = calendar.date(byAdding: .day, value: 1, to: cursor) else {
                break
            }
            cursor = next
        }

        return StreakState(currentStreak: current, longestStreak: longest, lastAchievedDate: lastAchievedDate)
    }
}
